### DISCUSSION TITLE
How to access docker server instance when tests are not in debug mode?

### DISCUSSION THREAD

**Discussion by bdw429s** [TechEmpower/FrameworkBenchmarks#6339] [General]
I'd like to be able to access a web-based monitor library in order to tune the performance of my framework WHILE the normal benchmarks are running on it.  

When I start the tests in debug mode with this command (I'm using Docker desktop):

```
$ docker run -it --rm --network tfb -v /var/run/docker.sock:/var/run/docker.sock -v /D_DRIVE/docs/GitHub/FrameworkBenchmarks:/FrameworkBenchmarks techempower/tfb --mode debug --test cfml
```
I am able to hit my site the browser:
![image](https://user-images.githubusercontent.com/584009/106964940-8abd6780-6708-11eb-8496-f5a463214f75.png)

And when I run the tests normally with this command:
```
$ docker run -it --rm --network tfb -v /var/run/docker.sock:/var/run/docker.sock -v /D_DRIVE/docs/GitHub/FrameworkBenchmarks:/FrameworkBenchmarks techempower/tfb --test cfml
```

I get this in the browser (while the cfml container is up and the tests are running):

![image](https://user-images.githubusercontent.com/584009/106965007-a294eb80-6708-11eb-8108-fa5c70747d8d.png)

@msmith-techempower said on the Google group that 

> The current toolset should not discern between benchmark and debug modes

but that's not the behavior I'm experiencing.

**Comment by bdw429s** [comment]
Any insight on this?

**Reply by msmith-techempower** [reply]
When you are running a benchmark (which should take a few minutes) what is the output from `docker ps`?

**Reply by bdw429s** [reply]
Thanks for the reply @msmith-techempower   When I run the tests normally (and the site is not accessible externally), `docker ps` shows this

```
root@default:/home/docker# docker ps
CONTAINER ID        IMAGE                       COMMAND                  CREATED             STATUS                    PORTS                NAMES
055a5e250faf        techempower/tfb.test.cfml   "/bin/sh -c $BUILD_D…"   36 seconds ago      Up 35 seconds (healthy)   8080/tcp, 8443/tcp   tfb-server
5c6b8e0a619b        techempower/postgres        "postgres"               38 seconds ago      Up 38 seconds                                  tfb-database
b377d4f6a3c2        techempower/tfb             "python /FrameworkBe…"   45 seconds ago      Up 44 seconds                                  elated_easley
```

However, when I run the test in debug mode, (and the pages are accessible externally), `docker ps` shows this:

```
root@default:/home/docker# docker ps
CONTAINER ID        IMAGE                       COMMAND                  CREATED             STATUS                             PORTS                              NAMES
c2e59de4fd86        techempower/tfb.test.cfml   "/bin/sh -c $BUILD_D…"   14 seconds ago      Up 14 seconds (health: starting)   0.0.0.0:8080->8080/tcp, 8443/tcp   tfb-server
b03557d9a9d5        techempower/postgres        "postgres"               17 seconds ago      Up 16 seconds                                                         tfb-database
a771bcdb3471        techempower/tfb             "python /FrameworkBe…"   22 seconds ago      Up 21 seconds                                                         amazing_hamilton
```

The first thing I notice there is that the port bindings appear to be handled differently in debug mode.  (`0.0.0.0:8080->8080/tcp` vs `8080/tcp` )

**Reply by msmith-techempower** [reply]
Ah, right. I had forgotten this little wrinkle - the legacy toolset only exposes ports during the `debug` mode. Verify/benchmark are expected to either run in `network-mode=host` or not accept external requests.

If you are trying to connect to your server instance while running verify/benchmark, you will need to add the `--network-mode host` flag.

**Reply by bdw429s** [reply]
So to be clear, are we saying that what I'm trying to accomplish is not possible?  The ` --network-mode host` flag just errors out

```
ConnectionError: HTTPConnectionPool(host='127.0.0.1', port=2375): Max retries exceeded with url...
```
It appears it's trying to connect to some running instance of the tests on my machine.  But I'm still wanting the benchmark suite to take care of running the tests inside the Docker containers like normal, I just want to be able to connect at the same time for performance tuning purposes.  Connecting my performance monitor doesn't do me any good in debug mode when nothing is running!

**Reply by bdw429s** [reply]
@msmith-techempower Did you have a chance to see my followup questions above?

**Reply by msmith-techempower** [reply]
Sorry, no. I have been out of the office since last week.

> So to be clear, are we saying that what I'm trying to accomplish is not possible?

I believe it *should* be possible, but maybe not if you are testing on Windows. On Unix, it should definitely be possible because that is tantamount to how we perform the real-world benchmarks.

**Reply by bdw429s** [reply]
Thanks for following up and I hope your time off was relaxing :)  

What can I do to help troubleshoot this?

**Reply by msmith-techempower** [reply]
Are you on Windows, Unix, or MacOS?

If you are on Unix, you will need to add two flags to your command - one for Docker and one for the toolset:

Docker: `--network host` 
toolset: `--network-mode host`

```
docker run -it --network host --rm --network tfb -v /var/run/docker.sock:/var/run/docker.sock -v /D_DRIVE/docs/GitHub/FrameworkBenchmarks:/FrameworkBenchmarks techempower/tfb --mode debug --test cfml --network-mode host
```

**Reply by bdw429s** [reply]
I am on Windows running Docker via a docker machine VM.  I will give that a try and report back.

**Reply by msmith-techempower** [reply]
Oof, okay well... since you're on Windows and the current toolset requires Docker-in-Docker, I am not sure that you can actually *run it* in `host` mode. I spent a *LONG* time trying, but I just could never get it working, which is one of the reasons I started writing the new toolset.

**Reply by bdw429s** [reply]
Right, so if I use just the docker `--network host` flag:
```
$ docker run -it --network host --rm -v /var/run/docker.sock:/var/run/docker.sock -v /D_DRIVE/docs/GitHub/FrameworkBenchmarks:/FrameworkBenchmarks te
chempower/tfb --test cfml
```
then I get the test seem to try to run, but the DB isn't accessible:
```
techempower/postgres:latest: Database was not ready after startup
```
and even though my CFML container starts, it isn't reachable.
```
Accessing URL http://tfb-server:8080/db.cfm:
Verifying test db for cfml caused an exception: HTTPConnectionPool(host='tfb-server', port=8080): Max retries exceeded with url: /db.cfm (Caused by Ne
wConnectionError('<urllib3.connection.HTTPConnection object at 0x7f0e56cff690>: Failed to establish a new connection: [Errno -3] Temporary failure in
name resolution',))
   FAIL for http://tfb-server:8080
     Server did not respond to request
```

And if I add the tool set flag of `--network-mode host` on the end, then regardless of my docker host flag, the tests immediately error out before they even start my CFML container with the following errors.  

```
================================================================================
Running Tests...
================================================================================
wrk: Docker build failed; terminating
wrk: Traceback (most recent call last):
wrk:   File "/FrameworkBenchmarks/toolset/utils/docker_helper.py", line 48, in __build
wrk:     decode=True
wrk:   File "/usr/local/lib/python2.7/dist-packages/docker/api/build.py", line 269, in build
wrk:     timeout=timeout,
wrk:   File "/usr/local/lib/python2.7/dist-packages/docker/utils/decorators.py", line 46, in inner
wrk:     return f(self, *args, **kwargs)
wrk:   File "/usr/local/lib/python2.7/dist-packages/docker/api/client.py", line 226, in _post
wrk:     return self.post(url, **self._set_request_timeout(kwargs))
wrk:   File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 590, in post
wrk:     return self.request('POST', url, data=data, json=json, **kwargs)
wrk:   File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 542, in request
wrk:     resp = self.send(prep, **send_kwargs)
wrk:   File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 655, in send
wrk:     r = adapter.send(request, **kwargs)
wrk:   File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 516, in send
wrk:     raise ConnectionError(e, request=request)
wrk: ConnectionError: HTTPConnectionPool(host='127.0.0.1', port=2375): Max retries exceeded with url: /v1.35/build?q=False&pull=True&t=techempower%2Ft
fb.wrk&nocache=False&forcerm=True&rm=False&dockerfile=wrk.dockerfile (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff
17012cd10>: Failed to establish a new connection: [Errno 111] Connection refused',))
wrk: Build time: 0s
A fatal error has occurred
Traceback (most recent call last):
  File "/FrameworkBenchmarks/toolset/run-tests.py", line 244, in main
    any_failed = benchmarker.run()
  File "/FrameworkBenchmarks/toolset/benchmark/benchmarker.py", line 57, in run
    self.docker_helper.build_wrk()
  File "/FrameworkBenchmarks/toolset/utils/docker_helper.py", line 364, in build_wrk
    tag="techempower/tfb.wrk")
  File "/FrameworkBenchmarks/toolset/utils/docker_helper.py", line 48, in __build
    decode=True
  File "/usr/local/lib/python2.7/dist-packages/docker/api/build.py", line 269, in build
    timeout=timeout,
  File "/usr/local/lib/python2.7/dist-packages/docker/utils/decorators.py", line 46, in inner
    return f(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/docker/api/client.py", line 226, in _post
    return self.post(url, **self._set_request_timeout(kwargs))
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 590, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
ConnectionError: HTTPConnectionPool(host='127.0.0.1', port=2375): Max retries exceeded with url: /v1.35/build?q=False&pull=True&t=techempower%2Ftfb.wr
k&nocache=False&forcerm=True&rm=False&dockerfile=wrk.dockerfile (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff17012
cd10>: Failed to establish a new connection: [Errno 111] Connection refused',))
Shutting down (may take a moment)
```

**Reply by msmith-techempower** [reply]
> And if I add the tool set flag of `--network-mode host` on the end, then regardless of my docker host flag, the tests immediately error out before they even start my CFML container with the following errors.

Right, this is expected on Windows. 

> then I get the test seem to try to run, but the DB isn't accessible:
> ```
> techempower/postgres:latest: Database was not ready after startup
> ```

Right, this is also expected when you only have `--network host` set on Docker. Essentially, that is telling Docker to do a few things, but largely it puts the started containers (application and database) on the host network. However, because you are missing the `--network-mode host` on the toolset, it is expecting the hostname `tfb-database` to resolve, but it doesn't because your host network knows nothing about that.

**Reply by bdw429s** [reply]
Thanks for trying to help.  It sounds like the final take away from this is that it's not possible.  I don't know much about the toolset, but that's a shame since the services are accessible in debug mode so it does seem like it should be possible.  I'll look forward to a future update of the toolset that allows this and in the mean time I suppose I'll have to use a JMeter test to try and simulate what the toolset is doing in debug mode.
